### PR TITLE
bolt12: add Invoice codec and structural validators

### PR DESCRIPTION
**Pull request by bitromortac**
Based on #10832, part of #10736.

Adds the BOLT 12 `Invoice` message struct with `PureTLVMessage`-based `Encode/Decode`, alongside the structural `ValidateInvoiceRead/Write` and cross-message `ValidateInvoiceAgainstRequest` validators.

Additionally, this PR injects known feature-bit catalogues into the `Offer` and `InvoiceRequest` read-side validators to enable must-understand capability checks at runtime, because features are defined outside of bolt12's boundaries.

### CODE REVIEW

File: bolt12/validate.go [hunk 1]
```diff
@@ -356,14 +423,6 @@ func ValidateInvoiceRequestWrite(ir *InvoiceRequest) error {
...
 
-	// - if it supports bolt12 invoice request features:
-	//   - MUST set invreq_features.features to the bitmap of features.
-	// We only reject unknown even bits here; advertising a feature is the
-	// caller's decision.
-	if err := checkFeatures(ir.InvreqFeatures); err != nil {
-		return err
-	}
-
 	// check UTF-8 constraints and BIP 353
```
Comment (bitromortac) [inline bolt12/validate.go:433]: I removed feature bit validation from writer validation as otherwise we'd need to pass in known features. I think it should be ok to rely on ourselves to set feature bits correctly.

File: bolt12/validate.go [hunk 2]
```diff
@@ -461,7 +520,8 @@ func getInvreqChain(ir *InvoiceRequest) [32]byte {
 // Invoice message (see the TODO at the end of this function). Until then, a
 // caller wiring this into a handler MUST verify the signature itself.
 func ValidateInvoiceRequestRead(ir *InvoiceRequest,
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply bolt12/validate.go:529]: Yes, callers were updated, so not an issue.

File: bolt12/validate.go [hunk 3]
```diff
@@ -1114,3 +1173,485 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+		return ErrBlindedPayMismatch
+	}
+
+	// - For each invoice_blindedpay.payinfo:
+	//   - MUST NOT use the corresponding invoice_paths.path if
+	//     payinfo.features has any unknown even bits set.
+	//   - MUST reject the invoice if this leaves no usable paths.
+	var usablePaths int
+	for i := range bp.Infos {
+		fv := bp.Infos[i].Features
```
Comment (bitromortac) [inline bolt12/validate.go]: Have added the docstring and signature validation will come later, in addition to checking against the blinded path.

File: bolt12/subtypes.go [hunk 4]
```diff
@@ -85,3 +93,296 @@ func decodeChainsRecord(r io.Reader, val any, _ *[8]byte, l uint64) error {
...
+	// WARNING: RawFeatureVector re-encodes to minimal length, so setting
+	// non-minimal feature bytes (trailing zeros) yields different wire
+	// bytes than were read and invalidates the invoice signature.
+	Features lnwire.RawFeatureVector
+}
+
+// BlindedPayInfos holds a list of BlindedPayInfo entries for the
+// invoice_blindedpay field.
+type BlindedPayInfos struct {
+	Infos []BlindedPayInfo
```
Comment (bitromortac) [inline bolt12/subtypes.go:161]: This is a known issue as documented. I favored using the internal data structure for tlv mechanics over keeping the byte-wise representation. Added an error path to detect this explicitly.

File: bolt12/subtypes.go [hunk 5]
```diff
@@ -85,3 +93,296 @@ func decodeChainsRecord(r io.Reader, val any, _ *[8]byte, l uint64) error {
...
+		binary.BigEndian.PutUint16(buf[:2], uint16(flen))
+		if _, err := w.Write(buf[:2]); err != nil {
+			return err
+		}
+		if err := info.Features.EncodeBase256(w); err != nil {
+			return err
+		}
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply bolt12/subtypes.go:248]: Addressed.

File: bolt12/validate.go [hunk 6]
```diff
@@ -1114,3 +1173,485 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+	// fields live on the invoice). NOT CHECKED HERE: the byte-for-byte
+	// field mirror and the invreq_amount == invoice_amount rule are
+	// enforced by ValidateInvoiceAgainstRequest once the invoice is paired
+	// with its request; the offer_paths blinded_node_id case needs the
+	// arrival path and stays with the caller.
+	if err := checkInvoiceNodeID(inv); err != nil {
+		return err
+	}
+
+	// - MUST reject the invoice if signature is not a valid signature using
```
Comment (bitromortac) [inline bolt12/validate.go:1724]: Yes, documented explicitly. Imo validation should only read and not modify in this case, so this has to be checked by the pathfinding system later.

File: bolt12/subtypes.go [hunk 7]
```diff
@@ -85,3 +117,305 @@ func decodeChainsRecord(r io.Reader, val any, _ *[8]byte, l uint64) error {
...
+	CltvExpiryDelta           uint16
+	HtlcMinimumMsat           uint64
+	HtlcMaximumMsat           uint64
+
+	// Features is the relay feature bitmap for this blinded path, typed for
+	// consistency with the other BOLT 12 feature fields.
+	//
+	// WARNING: RawFeatureVector re-encodes to minimal length, so setting
+	// non-minimal feature bytes (trailing zeros) yields different wire
+	// bytes than were read and invalidates the invoice signature.
```
Comment (erickcestari) [inline bolt12/subtypes.go:154]: I believe it would be worth a change in the spec to only allow minimal encoded `features` in all lightning types.

This would be similar to what is being done with `channel_announcement` and `node_announcement`:
https://github.com/lightning/bolts/pull/1341#pullrequestreview-4402013432

File: docs/release-notes/release-notes-0.22.0.md [hunk 8]
```diff
@@ -110,6 +99,13 @@
   in `SubscribeOnionMessages`, ensuring a nil reply path remains nil in the
   RPC response rather than being emitted as an empty struct.
 
+* [BOLT 12 invoice
+  codec](https://github.com/lightningnetwork/lnd/pull/10999): add the
```
Comment (erickcestari) [inline docs/release-notes/release-notes-0.22.0.md]: nit:
```suggestion
  codec](https://github.com/lightningnetwork/lnd/pull/10941): add the
```

File: bolt12/invoice.go [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+	]
+
+	// InvoiceFeatures are the features of the invoice.
+	InvoiceFeatures tlv.OptionalRecordT[
+		tlv.TlvType174, lnwire.RawFeatureVector,
+	]
+
+	// InvoiceNodeID is the public key of the recipient node, used to verify
+	// the signature.
+	InvoiceNodeID tlv.OptionalRecordT[tlv.TlvType176, [33]byte]
```
Comment (erickcestari) [inline bolt12/invoice.go]: nit:
```suggestion
	InvoiceNodeID tlv.OptionalRecordT[tlv.TlvType176, *btcec.PublicKey]
```

File: bolt12/validate.go [hunk 10]
```diff
@@ -1114,3 +1180,494 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+		fv := bp.Infos[i].Features
+		wrapped := lnwire.NewFeatureVector(&fv, knownBlindedFeatures)
+		if len(wrapped.UnknownRequiredFeatures()) == 0 {
+			usablePaths++
+		}
+	}
+
+	if usablePaths == 0 {
+		return ErrNoUsablePaths
+	}
```
Comment (erickcestari) [inline bolt12/validate.go:1703]: nit: We could have a similar method of `UsableFallbackAddresses`, but for the BlindedPaths fields.

```go
// UsablePath pairs a blinded path with its payment parameters, as returned by
// UsablePaths after the BOLT 12 reader's feature filter has been applied.
type UsablePath struct {
	// Path is the blinded path to the recipient.
	Path lnwire.BlindedPath

	// PayInfo is the blinded_payinfo for Path.
	PayInfo BlindedPayInfo
}

// UsablePaths returns the invoice_paths entries a payer may use, each paired
// with its blinded_payinfo, after applying the BOLT 12 reader rule that a path
// MUST NOT be used when its payinfo.features has unknown required (even) bits
// set. knownBlindedFeatures names the feature bits the reader understands.
//
// The result is empty when invoice_paths or invoice_blindedpay is absent, or
// when the two lists differ in length; ValidateInvoiceRead rejects those cases
// separately, so a caller that validates first can treat an empty result as
// "no usable paths".
func (inv *Invoice) UsablePaths(
	knownBlindedFeatures map[lnwire.FeatureBit]string) []UsablePath {

	paths := inv.InvoicePaths.ValOpt().UnwrapOr(lnwire.BlindedPaths{})
	bp := inv.InvoiceBlindedPay.ValOpt().UnwrapOr(BlindedPayInfos{})

	// Entries pair by index; a length mismatch is rejected upstream by
	// ValidateInvoiceRead, so guard here to stay in bounds.
	if len(paths.Paths) != len(bp.Infos) {
		return nil
	}

	var usable []UsablePath
	for i := range bp.Infos {
		// MUST NOT use the path if payinfo.features has any unknown even
		// bits set.
		fv := bp.Infos[i].Features
		wrapped := lnwire.NewFeatureVector(&fv, knownBlindedFeatures)
		if len(wrapped.UnknownRequiredFeatures()) > 0 {
			continue
		}

		usable = append(usable, UsablePath{
			Path:    paths.Paths[i],
			PayInfo: bp.Infos[i],
		})
	}

	return usable
}
```

Then:

```suggestion
	// - For each invoice_blindedpay.payinfo:
	//   - MUST NOT use the corresponding invoice_paths.path if
	//     payinfo.features has any unknown even bits set.
	//   - MUST reject the invoice if this leaves no usable paths.
	// UsablePaths applies that filter; a caller selecting a path downstream
	// should use it rather than the unfiltered invoice_paths.
	if len(inv.UsablePaths(knownBlindedFeatures)) == 0 {
		return ErrNoUsablePaths
	}
```

**Review by erickcestari — APPROVED** [review]
LGTM! Nice work!🏅

There are only some nits that are non-blocking from my side.

**Comment by saubyk** [comment]
/gateway dismiss F1 per https://github.com/lightningnetwork/lnd/pull/10941#discussion_r3506473767

**Comment by saubyk** [comment]
/gateway dismiss F6 per https://github.com/lightningnetwork/lnd/pull/10941#discussion_r3506508097

**Comment by saubyk** [comment]
/gateway dismiss F3 per https://github.com/lightningnetwork/lnd/pull/10941#discussion_r3506493905

**Comment by saubyk** [comment]
/gateway dismiss F4 per https://github.com/lightningnetwork/lnd/pull/10941#discussion_r3506500372

File: bolt12/validate.go [hunk 11]
```diff
@@ -461,7 +527,8 @@ func getInvreqChain(ir *InvoiceRequest) [32]byte {
 // Invoice message (see the TODO at the end of this function). Until then, a
 // caller wiring this into a handler MUST verify the signature itself.
 func ValidateInvoiceRequestRead(ir *InvoiceRequest,
-	activeChain [32]byte) error {
+	activeChain [32]byte,
```
Comment (bitromortac) [inline bolt12/validate.go:530]: This is re-flag, see other comment.

File: bolt12/validate.go [hunk 12]
```diff
@@ -1114,3 +1180,494 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+	if err := fn.MapOptionZ(inv.InvreqPayerID.ValOpt(),
+		func(pk *btcec.PublicKey) error {
+			if pk == nil {
+				return fmt.Errorf("%w: invreq_payer_id",
+					ErrNilPublicKey)
+			}
+
+			return nil
+		}); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply bolt12/validate.go:1357]: Good catch. I changed the comparator from greater than or equal to a strict greater than, so the exact expiry second is still accepted. That matches the reader spec and the behavior of ValidateOfferRead. I also added boundary tests covering the exact second and one second past it.

File: bolt12/validate.go [hunk 13]
```diff
@@ -1114,3 +1180,494 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+
+	if len(reqFields) > 0 {
+		return fmt.Errorf("%w: invoice is missing %d fields from "+
+			"request", ErrInvoiceMismatch, len(reqFields))
+	}
+
+	// Spec MUST: if invreq_amount (type 82) is present, invoice_amount
+	// (type 170) must equal it. The byte-mirror loop cannot relate fields
+	// with differing type numbers.
+	return fn.MapOptionZ(
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply bolt12/validate.go]: Agreed. When invreq_amount is absent I now enforce the native lower bound of offer_amount times invreq_quantity with an overflow guard, mirroring the request side check. The offer_currency case needs a caller supplied exchange rate, so it stays delegated and I documented that in the function contract. I added table tests for below, at, and above the bound plus the currency delegation skip.

File: bolt12/validate.go [hunk 14]
```diff
@@ -1114,3 +1180,494 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+	//     chain.
+	// invreq_chain defaults to bitcoin mainnet when absent. activeChain is
+	// the chain the reader supports.
+	chain := inv.InvreqChain.ValOpt().UnwrapOr(bitcoinMainnetGenesisHash)
+	if chain != activeChain {
+		return ErrUnsupportedChain
+	}
+
+	// - if invoice_features contains unknown odd bits that are non-zero:
+	//   - MUST ignore the bit.
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply bolt12/validate.go:1635]: Makes sense. I grouped the two catalogues into a named InvoiceFeatureCatalogues struct with Invoice and Blinded fields, so the two same typed maps can no longer be swapped positionally. The in package callers and tests are updated to match.

File: docs/release-notes/release-notes-0.22.0.md (hunk 8 above)
Comment (bitromortac) [reply docs/release-notes/release-notes-0.22.0.md]: Thanks, fixed the link to point at 10941.

File: bolt12/validate.go (hunk 10 above)
Comment (bitromortac) [reply bolt12/validate.go:1703]: Nice suggestion. I added Invoice.UsablePaths, which returns each usable path paired with its blinded_payinfo, and ValidateInvoiceRead now uses it in place of the ad hoc counting loop. I added a test for it as well.

File: bolt12/invoice.go (hunk 9 above)
Comment (bitromortac) [reply bolt12/invoice.go]: Good call, done. invoice_node_id is now a *btcec.PublicKey to match offer_issuer_id and invreq_payer_id. I added a present but nil guard in both the writer and reader validators, and the comparison against offer_issuer_id now uses IsEqual on the parsed keys. I also added rejection tests for the nil case on both validators.

**Comment by bitromortac** [comment]
Thanks for you review @erickcestari, addressed your comments and also the ones for lightninglabs-gateway. @vctt94 @Abdulkbk your reviews are highly welcome, if you find the time (I can't request reviews from you on github).

**Comment by saubyk** [comment]
/gateway dismiss F2 per https://github.com/lightningnetwork/lnd/pull/10941#discussion_r3536961683

File: bolt12/subtypes.go [hunk 15]
```diff
@@ -85,3 +117,305 @@ func decodeChainsRecord(r io.Reader, val any, _ *[8]byte, l uint64) error {
...
+
+		binary.BigEndian.PutUint64(buf[:8], info.HtlcMaximumMsat)
+		if _, err := w.Write(buf[:8]); err != nil {
+			return err
+		}
+
+		// flen is a u16, so guard the cast before framing the minimal
+		// feature bytes, mirroring encodeFallbackAddrs.
+		flen := info.Features.SerializeSize()
+		if flen > math.MaxUint16 {
```
[the remark this answers is not in the document]
Comment (bitromortac) [reply bolt12/subtypes.go:234]: `decodeBlindedPayInfos` now rejects an inverted range where `htlc_minimum_msat` exceeds `htlc_maximum_msat` with a typed `ErrInvalidHtlcRange`, right at decode so the invariant holds for every downstream consumer rather than being re-derived per call site. This mirrors the non-minimal-features guard already on the struct, and there is a decode test for the inverted case.

File: bolt12/subtypes.go [hunk 16]
```diff
@@ -85,3 +123,313 @@ func decodeChainsRecord(r io.Reader, val any, _ *[8]byte, l uint64) error {
 
 	return nil
 }
+
+// BlindedPayInfo holds the payment parameters for a blinded path, corresponding
+// to the blinded_payinfo subtype.
+type BlindedPayInfo struct {
```
Comment (Abdulkbk) [inline bolt12/subtypes.go:129]: nit: can add godoc for the remaining exported fields in the struct

File: bolt12/subtypes_test.go [hunk 17]
```diff
@@ -152,3 +154,262 @@ func TestChainsRecordRoundTrip(t *testing.T) {
...
+		new([8]byte), uint64(len(encoded)),
+	)
+	require.NoError(t, err)
+	require.Equal(t, infos.Infos, decoded.Infos)
+}
+
+// TestDecodeBlindedPayInfosRejectsTruncated covers truncation before the fixed
+// fields and before the declared features payload. Each must fail rather than
+// yield a partial BlindedPayInfos with corrupt entries.
+func TestDecodeBlindedPayInfosRejectsTruncated(t *testing.T) { //
```
Comment (Abdulkbk) [inline bolt12/subtypes_test.go]: empty comment `//` leftover at the end.

File: bolt12/invoice.go [hunk 18]
```diff
@@ -0,0 +1,426 @@
...
+func (inv *Invoice) UsableFallbackAddresses() []FallbackAddress {
+	// Unwrap the optional up front so the filtering loop stays flat; a nil
+	// Addrs slice ranges as empty.
+	fallbacks := inv.InvoiceFallbacks.ValOpt().UnwrapOr(FallbackAddresses{})
+
+	var addrs []FallbackAddress
+	for _, a := range fallbacks.Addrs {
+		// MUST ignore any fallback_address for which version is greater
+		// than 16.
+		if a.Version > 16 {
```
Comment (Abdulkbk) [inline bolt12/invoice.go]: nits: The nums: 16, 2, and 40 in this method are specs nums. Consider using named constants

**Review by Abdulkbk — APPROVED** [review]
LGTM 🚢

File: bolt12/subtypes_test.go (hunk 17 above)
Comment (bitromortac) [reply bolt12/subtypes_test.go]: Thanks for catching this

File: bolt12/invoice.go (hunk 18 above)
Comment (bitromortac) [reply bolt12/invoice.go]: They are extracted now :+1:

File: bolt12/validate.go [hunk 19]
```diff
@@ -1114,3 +1180,572 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+	// checkFeatures enforces those invoice_features bit rules below.
+	//
+	// Separately, BOLT 1 makes unknown even TLV types must-understand, so
+	// reject those here over the decoded type set. Unlike the
+	// invoice_request reader, the invoice reader defines no out-of-range
+	// type rejection, so unknown odd types are simply ignored ("it's ok to
+	// be odd"). The signature range (240-1000) is exempt for the same
+	// reason, matching the invoice_request reader and the Merkle path.
+	for _, t := range sortedTypes(inv.decodedTLVs) {
+		if bolt12InUnsignedRange(t) {
```
Comment (vctt94) [inline bolt12/validate.go:1647]: Non-blocking: should this match the invoice_request pattern and only exempt the signature range from range checks, not from unknown-even checks?

Right now an unknown even TLV in the 240-1000 signature range would be skipped before ErrUnknownEvenType is applied. ValidateInvoiceRequestRead still rejects unknown even types inside that range, which seems like the safer/more consistent behavior.

**Review by vctt94 — APPROVED** [review]
LGTM! Left only a small nit.

great job!

File: bolt12/subtypes.go (hunk 16 above)
Comment (bitromortac) [reply bolt12/subtypes.go:129]: Added godoc for the remaining exported `BlindedPayInfo` fields (`FeeBaseMsat`, `FeeProportionalMillionths`, `CltvExpiryDelta`, `HtlcMinimumMsat`, `HtlcMaximumMsat`), so the whole struct is documented now.

File: bolt12/validate.go (hunk 19 above)
Comment (ziggie1984) [reply bolt12/validate.go:1647]: +1 on this

File: bolt12/validate.go [hunk 20]
```diff
@@ -1114,3 +1180,572 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+		return nil
+	}
+
+	offerAmt := uint64(inv.OfferAmount.ValOpt().UnwrapOr(0))
+	qty := uint64(inv.InvreqQuantity.ValOpt().UnwrapOr(1))
+	invAmt := uint64(inv.InvoiceAmount.ValOpt().UnwrapOr(0))
+
+	// Guard against overflow of offer_amount * quantity.
+	hi, expectedAmt := bits.Mul64(offerAmt, qty)
+	if hi != 0 {
```
Comment (ziggie1984) [inline bolt12/validate.go:1529]: nit: the overflow guard here (`hi != 0`) has no test coverage — it looks like the only fully-uncovered branch in the new invoice validation. The request-side twin `checkInvreqAmountMeetsOffer` is exercised by `TestValidateInvoiceRequestAmountOverflow`, but the invoice side isn't. A case in `TestValidateInvoiceAgainstRequestOfferAmount` with `offer_amount = math.MaxUint64`, `invreq_quantity = 2`, and `invreq_amount` absent, asserting `ErrAmountBelowExpected`, would close it.

File: bolt12/subtypes_test.go [hunk 21]
```diff
@@ -152,3 +154,262 @@ func TestChainsRecordRoundTrip(t *testing.T) {
...
+			t.Parallel()
+
+			var bp BlindedPayInfos
+			err := decodeBlindedPayInfos(
+				bytes.NewReader(tc.data), &bp, new([8]byte),
+				tc.declLen,
+			)
+			require.Error(t, err)
+			require.Contains(t, err.Error(), tc.errSubstr)
+		})
```
Comment (ziggie1984) [inline bolt12/subtypes_test.go:324]: nit: these cases assert on the error string (`require.Contains(err.Error(), ...)`) even though the decoder wraps `%w` sentinels — `ErrTooManyBlindedPayInfos`, `ErrNonMinimalFeatures`, `ErrInvalidHtlcRange` (and `ErrTooManyFallbackAddrs` in the fallback test below). `TestDecodeChainsRecord` already uses `require.ErrorIs(err, ErrTooManyChains)`; switching these to `require.ErrorIs` makes them robust to error-text drift.

File: bolt12/invoice.go [hunk 22]
```diff
@@ -0,0 +1,440 @@
...
+
+			continue
+		}
+
+		// MUST ignore any fallback_address for which address does not
+		// meet known requirements for the given version. NOT enforced
+		// here: the per-version witness-program check needs on-chain
+		// address rules above this codec, so a caller dispatching
+		// on-chain MUST apply it.
+		addrs = append(addrs, a)
```
Comment (GeorgeTsagk) [inline bolt12/invoice.go:189]: the current checks still leave room for invalid addresses to be included here. Do we want to tighten up the requirements?

File: bolt12/validate.go [hunk 23]
```diff
@@ -359,8 +359,10 @@ func ValidateInvoiceRequestWrite(ir *InvoiceRequest) error {
 	// - if it supports bolt12 invoice request features:
 	//   - MUST set invreq_features.features to the bitmap of features.
 	// We only reject unknown even bits here; advertising a feature is the
-	// caller's decision.
-	if err := checkFeatures(ir.InvreqFeatures); err != nil {
+	// caller's decision. Since the writer lacks a catalogue in scope, we
```
Comment (ziggie1984) [inline bolt12/validate.go:362]: what do you mean by lacking a catalogue, you mean there are currently no even features which should be supported the bolt12 validation ?

File: bolt12/validate.go [hunk 24]
```diff
@@ -359,8 +359,10 @@ func ValidateInvoiceRequestWrite(ir *InvoiceRequest) error {
...
+	// pass nil for the catalogue, treating all even feature bits as
+	// unknown.
+	if err := checkFeatures(ir.InvreqFeatures, nil); err != nil {
```
Comment (ziggie1984) [inline bolt12/validate.go:365]: hmm the final version however has:

```
if err := checkFeatures(ir.InvreqFeatures, knownFeatures); err != nil {
```

?

File: bolt12/validate.go [hunk 25]
```diff
@@ -1029,15 +1033,13 @@ func checkISO4217[T tlv.TlvType](opt tlv.OptionalRecordT[T, tlv.Blob]) error {
 
 // checkFeatures rejects any unknown even (must-understand) feature bit.
 func checkFeatures[T tlv.TlvType](
-	opt tlv.OptionalRecordT[T, lnwire.RawFeatureVector]) error {
+	opt tlv.OptionalRecordT[T, lnwire.RawFeatureVector],
+	known map[lnwire.FeatureBit]string) error {
```
Comment (ziggie1984) [inline bolt12/validate.go:1037]: Nit: call it `knownFeatures`

**Review by GeorgeTsagk — APPROVED** [review]
LGTM

File: bolt12/invoice.go (hunk 22 above)
Comment (ziggie1984) [reply bolt12/invoice.go:189]: yes we can make those address check more strict for example the taproot pubkey script

File: bolt12/invoice.go [hunk 26]
```diff
@@ -0,0 +1,436 @@
...
+	if len(paths.Paths) != len(bp.Infos) {
+		return nil
+	}
+
+	var usable []UsablePath
+	for i := range bp.Infos {
+		// MUST NOT use the path if payinfo.features has any unknown
+		// even bits set.
+		fv := bp.Infos[i].Features
+		wrapped := lnwire.NewFeatureVector(&fv, knownBlindedFeatures)
```
Comment (ziggie1984) [inline bolt12/invoice.go:231]: wondering why we do not use the checkFeature function here ?

File: bolt12/validate.go [hunk 27]
```diff
@@ -358,13 +425,10 @@ func ValidateInvoiceRequestWrite(ir *InvoiceRequest) error {
...
 	// - if it supports bolt12 invoice request features:
 	//   - MUST set invreq_features.features to the bitmap of features.
-	// We only reject unknown even bits here; advertising a feature is the
-	// caller's decision. Since the writer lacks a catalogue in scope, we
-	// pass nil for the catalogue, treating all even feature bits as
-	// unknown.
-	if err := checkFeatures(ir.InvreqFeatures, nil); err != nil {
-		return err
-	}
+	// We rely on the writer to set feature bits correctly as those are
```
Comment (ziggie1984) [inline bolt12/validate.go:428]: Nit: maybe prefix it with a `NOTE:` so it is easier when parsing the code here.

Comment (ziggie1984) [reply bolt12/validate.go:428]: I wonder how you decide when we do validate things on the writer side and when we don't, I think it is good to have a proper consistency and don't decide case by case, wdyt ?

File: bolt12/validate.go [hunk 28]
```diff
@@ -1116,3 +1180,572 @@ func checkPubKeyNotNil[T tlv.TlvType](
 		return nil
 	})
 }
+
+// checkInvoiceNodeID enforces the spec rule that, when offer_issuer_id is
+// present, invoice_node_id MUST equal it. Both fields live on the invoice, so
+// this is verifiable without the originating offer. The offer_paths branch
+// (invoice_node_id equals the final blinded_node_id on the arrival path) needs
```
Comment (ziggie1984) [inline bolt12/validate.go:1187]: could you clarify this a bit more, what is the blinded_node_id arrival path of an invoice ?

File: bolt12/validate.go [hunk 29]
```diff
@@ -1116,3 +1180,572 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+// caller context and is not checked here. A present-but-nil offer_issuer_id or
+// invoice_node_id is rejected separately as ErrNilPublicKey, so a nil here is
+// treated as absent.
+func checkInvoiceNodeID(inv *Invoice) error {
+	// A present-but-nil offer_issuer_id is rejected separately as
```
Comment (ziggie1984) [inline bolt12/validate.go:1192]: you mean it is rejected somewhere else in the code and this can bascially never happen if we check it here ?

File: bolt12/validate.go [hunk 30]
```diff
@@ -1116,3 +1180,572 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+		}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// defaultInvoiceRelativeExpiry is the spec-defined fallback when an invoice
+// omits invoice_relative_expiry: two hours from creation.
+const defaultInvoiceRelativeExpiry uint32 = 7200
```
Comment (ziggie1984) [inline bolt12/validate.go:1376]: is this also accounted for in the blindedd path expiry (I think in the blinded path we check expiry by blocks so I wonder if this calculation determines the blinded path generation CLTV timeout as well then ?

File: bolt12/validate.go [hunk 31]
```diff
@@ -1116,3 +1180,572 @@ func checkPubKeyNotNil[T tlv.TlvType](
...
+// and the same ExtraSignedFields entries produce byte-identical encodings for
+// any matching type. That is the byte-mirror invariant the spec demands.
+//
+// The amount cross-check enforces the spec's authorized-range rule: when
+// invreq_amount is present, invoice_amount MUST equal it; otherwise the payer
+// relied on the offer's fixed amount, so invoice_amount MUST be at least
+// offer_amount * invreq_quantity for the native (bitcoin) case. The
+// offer_currency case needs a caller-supplied exchange rate and is delegated to
+// the caller.
+func ValidateInvoiceAgainstRequest(inv *Invoice, req *InvoiceRequest) error {
```
Comment (ziggie1984) [inline bolt12/validate.go:1450]: This makes sense, but I wonder why we do not have something similar comparing the invoice_request with the offer which also has mirrored fields ?

**Review by ziggie1984 — APPROVED** [review]
LGTM, I think we will see potential gaps only when the whole thing is implemented and we start creating Offers and Invoices in real time

Still take a look at my comments and decide by yourself if you want to address them here or in a followup PR, nothing really blocking this PR.
